src/map.cpp

```diff
@@ -238,95 +237,96 @@ bool Map::placeCreature(const Position& centerPos, Creature* creature, bool exte
 
 	int32_t index = 0;
 	uint32_t flags = 0;
-	Item* toItem = nullptr;
+	std::shared_ptr<Item> toItem = nullptr;
 
-	Cylinder* toCylinder = tile->queryDestination(index, *creature, &toItem, flags);
-	toCylinder->internalAddThing(creature);
+	auto toThing = tile->queryDestination(index, creature, toItem, flags);
+	toThing->internalAddThing(creature);
 
-	const Position& dest = toCylinder->getPosition();
+	const Position& dest = toThing->getPosition();
 	getQTNode(dest.x, dest.y)->addCreature(creature);
 	return true;
 }
 
-void Map::moveCreature(Creature& creature, Tile& newTile, bool forceTeleport /* = false*/)
+void Map::moveCreature(std::shared_ptr<Creature> creature, std::shared_ptr<Tile> newTile,
+                       bool forceTeleport /* = false*/)
 {
-	Tile& oldTile = *creature.getTile();
+	auto oldTile = creature->getTile();
 
 	// If the tile does not have the creature it means that the creature is ready for elimination, we skip the move.
-	if (!oldTile.hasCreature(&creature)) {
+	if (!oldTile->hasCreature(creature)) {
 		return;
 	}
 
-	Position oldPos = oldTile.getPosition();
-	Position newPos = newTile.getPosition();
+	Position oldPos = oldTile->getPosition();
+	Position newPos = newTile->getPosition();
 
-	bool teleport = forceTeleport || !newTile.getGround() || !oldPos.isInRange(newPos, 1, 1, 0);
+	bool teleport = forceTeleport || !newTile->getGround() || !oldPos.isInRange(newPos, 1, 1, 0);
 
 	SpectatorVec spectators, newPosSpectators;
 	getSpectators(spectators, oldPos, true);
 	getSpectators(newPosSpectators, newPos, true);
 	spectators.addSpectators(newPosSpectators);
 
 	std::vector<int32_t> oldStackPosVector;
-	for (Creature* spectator : spectators) {
-		if (Player* tmpPlayer = spectator->getPlayer()) {
-			if (tmpPlayer->canSeeCreature(&creature)) {
-				oldStackPosVector.push_back(oldTile.getClientIndexOfCreature(tmpPlayer, &creature));
+	for (const auto& spectator : spectators) {
+		if (auto tmpPlayer = spectator->getPlayer()) {
```

Comment:
Could this be `const auto`
```suggestion
		if (const auto tmpPlayer = spectator->getPlayer()) {
```